File: reversenodesinkgroup.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    
    class Bundle {
        public ListNode beg;
        public ListNode end;
        Bundle(ListNode beg, ListNode end){
            this.beg = beg;
            this.end = end;
        }
    }
    
    private Bundle reverseBet(Bundle b){
        
        if(b.beg==b.end) return b;
        
        ListNode pPre = null;
        ListNode p = b.beg;
        ListNode tmp;
        while(p!=b.end){
            tmp = p.next;
            p.next = pPre;
            pPre = p;
            p = tmp;
        }
        p.next = pPre;
        return new Bundle(b.end,b.beg);
    }
    
    public ListNode reverseKGroup(ListNode head, int k) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ListNode ans = head;
        ListNode p = head;
        ListNode pPre = new ListNode(0);
        pPre.next = head;
        Bundle lastBundle = new Bundle(pPre,pPre);
        int no = 0;
        while(p!=null){
            no++;
            if(no==k){
                no=0;
                ListNode pnext = p.next;
                Bundle kb = reverseBet(new Bundle(lastBundle.end.next,p));
                lastBundle.end.next = kb.beg;
                kb.end.next = pnext;
                lastBundle = kb;
                p = kb.end;
            }
            p = p.next;
        }
        return pPre.next;
    }
}

File: cpp/ReverseNodesinkGroup.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        ListNode* p = dummy;
        while(true) {
            bool haveValidKNodes = true;
            ListNode* q = p->next;
            for(int i=0;i<k;i++) {
                if(q==NULL) {
                    haveValidKNodes = false;
                    break;
                }
                q = q->next;
            }
            if(haveValidKNodes) {
                ListNode* n0 = p;
                ListNode* n1 = p->next;
                ListNode* n2 = p->next->next;
                ListNode* start = p->next;
                for(int i=0;i<k;i++) {
                    n1->next = n0;
                    n0 = n1;
                    n1 = n2;
                    if(n1!=NULL) {
                        n2 = n1->next;
                    }
                }
                start->next = n1;
                p->next = n0;
                p = start;
            } else{
                break;
            }
        }
        ListNode* res = dummy->next;
        delete dummy;
        return res;
    }
};
